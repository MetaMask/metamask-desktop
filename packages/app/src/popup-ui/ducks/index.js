import { combineReducers } from 'redux';
import { persistReducer, createMigrate } from 'redux-persist';
import appMigrations from '../../ui/migrations/app';
import appReducer from '../../ui/ducks/app/app';
import sendReducer from '../../../submodules/extension/ui/ducks/send/send';
import historyReducer from '../../../submodules/extension/ui/ducks/history/history';
import appStateReducer from '../../../submodules/extension/ui/ducks/app/app';
import confirmTransactionReducer from '../../../submodules/extension/ui/ducks/confirm-transaction/confirm-transaction.duck';
import metamaskReducer from '../../../submodules/extension/ui/ducks/metamask/metamask';
import domainsReducer from '../../../submodules/extension/ui/ducks/domains';
import gasReducer from '../../../submodules/extension/ui/ducks/gas/gas.duck';
import localeMessagesReducer from '../../../submodules/extension/ui/ducks/locale/locale';

const appPersistConfig = {
  key: 'app',
  storage: window.popupElectronBridge.appStore,
  migrate: createMigrate(appMigrations, { debug: false }),
  version: 0,
};

const persistedAppReducer = persistReducer(appPersistConfig, appReducer);

const rootReducer = combineReducers({
  app: persistedAppReducer,
  metamask: metamaskReducer,
  localeMessages: localeMessagesReducer,
  send: sendReducer,
  history: historyReducer,
  appState: appStateReducer,
  confirmTransaction: confirmTransactionReducer,
  DNS: domainsReducer,
  gas: gasReducer,
});

export default rootReducer;
