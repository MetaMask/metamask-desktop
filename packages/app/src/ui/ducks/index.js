import { combineReducers } from 'redux';
import { persistReducer, createMigrate } from 'redux-persist';

import pairStatusMigrations from '../migrations/pair-status';
import appMigrations from '../migrations/app';
import appReducer from './app/app';
import pairStatusReducer from './pair-status/pair-status';

const pairStatusPersistConfig = {
  key: 'pairStatus',
  storage: window.electronBridge.pairStatusStore,
  blacklist: ['connections', 'isWebSocketConnected', 'isDesktopPaired'],
  whitelist: ['isSuccessfulPairSeen', 'lastActivation'],
  migrate: createMigrate(pairStatusMigrations, { debug: false }),
  version: 0,
};
const persistedPairStatusReducer = persistReducer(
  pairStatusPersistConfig,
  pairStatusReducer,
);

const appPersistConfig = {
  key: 'app',
  storage: window.electronBridge.appStore,
  migrate: createMigrate(appMigrations, { debug: false }),
  version: 0,
};
const persistedAppReducer = persistReducer(appPersistConfig, appReducer);

const rootReducer = combineReducers({
  app: persistedAppReducer,
  pairStatus: persistedPairStatusReducer,
});

export default rootReducer;
