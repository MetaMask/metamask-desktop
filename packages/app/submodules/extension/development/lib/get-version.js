const { version: manifestVersion } = require('../../package.json');
const { BuildType } = require('./build-type');

/**
 * Get the current version of the MetaMask extension. The base manifest version
 * is modified according to the build type and version.
 *
 * The build version is needed because certain build types (such as beta) may
 * be released multiple times during the release process.
 *
 * @param {BuildType} buildType - The build type.
 * @param {number} buildVersion - The build version.
 * @returns {string} The MetaMask extension version.
 */
function getVersion(buildType, buildVersion) {
  return buildType === BuildType.main
    ? manifestVersion
    : `${manifestVersion}-${buildType}.${buildVersion}`;
}

module.exports = { getVersion };
