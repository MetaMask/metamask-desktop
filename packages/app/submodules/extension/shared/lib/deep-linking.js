export function openCustomProtocol(protocolLink) {
  return new Promise((resolve, reject) => {
    if (window?.navigator?.msLaunchUri) {
      window.navigator.msLaunchUri(protocolLink, resolve, () => {
        reject(new Error('Failed to open custom protocol link'));
      });
    } else {
      const timeoutId = window.setTimeout(function () {
        reject(new Error('Timeout opening custom protocol link'));
      }, 500);
      window.addEventListener('blur', function () {
        window.clearTimeout(timeoutId);
        resolve();
      });
      window.location = protocolLink;
    }
  });
}
