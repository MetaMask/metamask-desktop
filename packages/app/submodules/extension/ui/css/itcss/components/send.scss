.send-screen-wrapper {
  display: flex;
  flex-flow: column nowrap;
  z-index: 25;

  @include screen-sm-max {
    width: 100%;
    overflow-y: auto;
  }

  section {
    flex: 0 0 auto;
  }
}

.send-screen-card {
  background-color: var(--color-background-default);
  box-shadow: var(--shadow-size-xs) var(--color-shadow-default);
  padding: 46px 40.5px 26px;
  position: relative;
  align-items: center;
  display: flex;
  flex-flow: column nowrap;
  width: 498px;
  flex: 1 0 auto;

  @include screen-sm-max {
    top: 0;
    width: 100%;
    box-shadow: none;
    padding: 12px;
  }
}

/* Send Screen */

.send-screen section {
  margin: 4px 16px;
}

.send-screen input {
  @include H7;

  width: 100%;
}

.send-eth-icon {
  border-radius: 50%;
  width: 70px;
  height: 70px;
  border: 1px solid var(--color-border-muted);
  box-shadow: var(--shadow-size-xs) var(--color-shadow-default);
  position: absolute;
  top: -35px;
  z-index: 25;
  padding: 4px;
  background-color: var(--color-background-default);

  @include screen-sm-max {
    position: relative;
    top: 0;
  }
}

.send-screen-input-wrapper {
  width: 95%;
  position: relative;

  .fa-bolt {
    padding-right: 4px;
  }

  .large-input {
    @include Paragraph;

    border: 1px solid var(--color-border-default);
    border-radius: 4px;
    margin: 4px 0 20px;
  }

  .send-screen-gas-input {
    border: 1px solid transparent;
  }

  &__error-message {
    display: none;
  }

  &--error {
    input,
    .send-screen-gas-input {
      border-color: var(--color-error-default) !important;
    }

    .send-screen-input-wrapper__error-message {
      @include H7;

      display: block;
      position: absolute;
      bottom: 4px;
      left: 8px;
      color: var(--color-error-default);
    }
  }

  .send-screen-input-wrapper__error-message {
    @include H7;

    display: block;
    position: absolute;
    bottom: 4px;
    left: 8px;
    color: var(--color-error-default);
  }
}

.send-screen-input {
  width: 100%;
}

.send-screen-gas-input {
  @include Paragraph;

  width: 100%;
  height: 41px;
  border-radius: 3px;
  background-color: var(--color-background-default);
  border-width: 0;
  border-style: none;
  display: flex;
  justify-content: space-between;
  align-items: center;
  padding-left: 10px;
  padding-right: 12px;
  color: var(--color-text-alternative);
}

.send-screen-amount-labels {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
}

.send-screen-gas-labels {
  display: flex;
  flex-direction: row;
  justify-content: space-between;
}

.currency-toggle {
  &__item {
    color: var(--color-primary-default);
    cursor: pointer;

    &--selected {
      color: var(--color-text-default);
      cursor: default;
    }
  }
}

.send-screen-gas-input-customize {
  @include H7;

  color: var(--color-primary-default);
  cursor: pointer;
}

.gas-tooltip-close-area {
  position: fixed;
  top: 0;
  left: 0;
  z-index: 1000;
  width: 100%;
  height: 100%;
}

.gas-tooltip-arrow {
  height: 25px;
  width: 25px;
  z-index: 1200;
  background: var(--color-background-default);
  position: absolute;
  transform: rotate(45deg);
  left: 107px;
  top: 294px;
  box-shadow: var(--shadow-size-xs) var(--color-shadow-default);
}

.gas-tooltip {
  display: flex;
  justify-content: center;
}

.gas-tooltip-label {
  @include Paragraph;

  color: var(--color-text-alternative);
}

.gas-tooltip-header {
  padding-bottom: 12px;
}

.gas-tooltip-input-label {
  margin-bottom: 5px;
}

.gas-tooltip-input-label i {
  color: var(--color-icon-muted);
  margin-left: 6px;
}

.gas-tooltip-input-detail {
  @include H7;

  position: absolute;
  top: 4px;
  right: 26px;
  color: var(--color-text-muted);
}

.gas-tooltip-input-arrows {
  @include H6;

  position: absolute;
  top: 0;
  right: 4px;
  width: 17px;
  height: 28px;
  border: 1px solid var(--color-icon-muted);
  border-left: 0;
  display: flex;
  flex-direction: column;
  color: var(--color-text-muted);
  padding: 1px 4px;
  cursor: pointer;
}

.token-gas {
  &__amount {
    display: inline-block;
    margin-right: 4px;
  }

  &__symbol {
    display: inline-block;
  }
}

.send-screen {
  &__title {
    @include H4;

    color: var(--color-text-alternative);
  }

  &__subtitle {
    @include H6;

    margin: 10px 0 20px;
  }

  &__send-button,
  &__cancel-button {
    width: 163px;
    text-align: center;
  }

  &__send-button__disabled {
    opacity: 0.5;
    cursor: auto;
  }
}

.send-token {
  display: flex;
  flex-flow: column nowrap;
  z-index: 25;

  &__content {
    width: 498px;
    height: 605px;
    background-color: var(--color-background-default);
    box-shadow: var(--shadow-size-xs) var(--color-shadow-default);
    padding: 46px 40.5px 26px;
    position: relative;
    align-items: center;
    display: flex;
    flex-flow: column nowrap;
    flex: 1 0 auto;

    @include screen-sm-max {
      top: 0;
      width: 100%;
      box-shadow: none;
      padding: 12px;
    }
  }

  .identicon {
    position: absolute;
    top: -35px;
    z-index: 25;

    @include screen-sm-max {
      position: relative;
      top: 0;
      flex: 0 0 auto;
    }
  }

  &__title {
    @include H4;

    color: var(--color-text-alternative);
  }

  &__description,
  &__balance-text,
  &__token-symbol {
    @include H6;

    margin-top: 10px;
    text-align: center;
  }

  &__token-balance {
    @include H1;

    margin-top: 13px;

    .token-balance__amount {
      padding-right: 12px;
    }
  }

  &__button-group {
    display: flex;
    flex-flow: column nowrap;
    align-items: center;
    flex: 0 0 auto;

    @include screen-sm-max {
      margin-top: 24px;
    }

    button {
      width: 163px;
    }
  }
}

.confirm-send-token {
  &__hero-amount-wrapper {
    width: 100%;
  }
}

.send-v2 {
  &__container {
    width: 380px;
    border-radius: 8px;
    background-color: var(--color-background-default);
    box-shadow: var(--shadow-size-xs) var(--color-shadow-default);
    display: flex;
    flex-flow: column nowrap;
    z-index: 25;
    align-items: center;
    position: relative;

    @include screen-sm-max {
      width: 100%;
      top: 0;
      box-shadow: none;
      flex: 1 1 auto;
    }
  }

  &__send-header-icon-container {
    z-index: 25;

    @include screen-sm-max {
      position: relative;
      top: 0;
    }
  }

  &__send-header-icon {
    border-radius: 50%;
    width: 48px;
    height: 48px;
    border: 1px solid var(--color-border-muted);
    z-index: 25;
    padding: 4px;
    background-color: var(--color-background-default);
  }

  &__arrow-background {
    background-color: var(--color-background-default);
    height: 14px;
    width: 14px;
    position: absolute;
    top: 52px;
    left: 199px;
    border-radius: 50%;
    z-index: 100;

    @include screen-sm-max {
      top: 36px;
    }
  }

  &__header {
    height: 88px;
    width: 380px;
    background-color: var(--color-background-alternative);
    position: relative;
    display: flex;
    justify-content: center;
    align-items: center;

    @include screen-sm-max {
      height: 59px;
      width: 100vw;
    }
  }

  &__header-tip {
    height: 25px;
    width: 25px;
    background-color: var(--color-background-alternative);
    position: absolute;
    transform: rotate(45deg);
    left: 178px;
    top: 75px;

    @include screen-sm-max {
      top: 46px;
      left: 0;
      right: 0;
      margin: 0 auto;
    }
  }

  &__title {
    @include H3;

    color: var(--color-text-alternative);
    text-align: center;
    margin-top: 25px;
  }

  &__copy {
    @include H6;

    color: var(--color-text-alternative);
    text-align: center;
    margin-top: 10px;
    width: 287px;
  }

  &__error {
    @include H7;

    left: 8px;
    color: var(--color-error-default);
  }

  &__error-amount {
    margin-top: 5px;
  }

  &__warning {
    @include H7;

    left: 8px;
    color: var(--color-warning-default);
  }

  &__error-border {
    color: var(--color-error-default);
  }

  &__form {
    @include screen-sm-max {
      margin: 0;
      flex: 1 1 auto;
    }
  }

  &__form-header,
  &__form-header-copy {
    width: 100%;
    display: flex;
    flex-flow: column;
    align-items: center;
  }

  &__form-row {
    margin: 1rem 1rem 0;
    position: relative;
    display: flex;
    flex-flow: row;
    flex: 1 0 auto;
    justify-content: space-between;
    align-items: center;
  }

  &__form-field-container {
    display: flex;
    flex-direction: column;
    width: 277px;
  }

  &__form-field {
    flex: 1 1 auto;
    min-width: 0;

    .currency-display {
      color: var(--color-text-alternative);

      &__currency-symbol {
        color: var(--color-text-alternative);
      }

      &__converted-value,
      &__converted-currency {
        color: var(--color-text-alternative);
      }
    }

    .account-list-item {
      &__account-secondary-balance {
        color: var(--color-text-alternative);
      }
    }
  }

  &__form-label {
    @include Paragraph;

    color: var(--color-text-alternative);
    width: 95px;
    flex: 0 0 auto;
  }

  &__from-dropdown,
  &__asset-dropdown {
    @include H7;

    width: 100%;
    border: 1px solid var(--color-border-muted);
    border-radius: 4px;
    background-color: var(--color-background-default);
    color: var(--color-text-default);
    position: relative;

    &__close-area {
      position: fixed;
      top: 0;
      left: 0;
      z-index: 1000;
      width: 100%;
      height: 100%;
    }

    &__list {
      z-index: 1050;
      position: absolute;
      height: 220px;
      width: 100%;
      border: 1px solid var(--color-border-muted);
      border-radius: 4px;
      background-color: var(--color-background-default);
      box-shadow: var(--shadow-size-sm) var(--color-shadow-default);
      margin-top: 11px;
      margin-left: -1px;
      overflow-y: scroll;
    }
  }

  &__from-dropdown {
    height: 73px;
  }

  &__asset-dropdown {
    height: 54px;
    border: none;
    position: relative;

    &__single-asset {
      display: flex;
      flex-flow: row nowrap;
      align-items: center;
      padding: 8px 8px;
    }

    &__caret {
      color: var(--color-icon-default);
      padding: 0 10px;
    }

    &__asset {
      display: flex;
      flex-flow: row nowrap;
      align-items: center;
      padding: 0 8px;
      cursor: pointer;
    }

    &__asset-icon {
      .identicon {
        border: 1px solid var(--color-border-muted);
      }
    }

    &__asset-data {
      display: flex;
      flex-flow: column nowrap;
      margin-left: 8px;
      flex-grow: 1;
    }

    @include screen-sm-max {
      &__asset-data {
        width: 60%;
      }
    }

    &__symbol {
      @include Paragraph;

      margin-bottom: 2px;
      overflow: hidden;
      text-overflow: ellipsis;
      white-space: nowrap;
    }

    &__name {
      @include H7;

      display: flex;
      flex-flow: row nowrap;

      &__label {
        margin-right: 0.25rem;
        min-width: 52px;
      }
    }

    &__close-area {
      z-index: 2000;
    }

    &__list {
      z-index: 2050;
      position: absolute;
      height: 220px;
      width: 100%;
      border: 1px solid var(--color-border-muted);
      border-radius: 4px;
      background-color: var(--color-background-default);
      box-shadow: var(--shadow-size-sm) var(--color-shadow-default);
      top: 65px;
      left: 0;
      box-sizing: content-box;
      overflow-y: scroll;
      margin-top: 0;
      padding: 4px 0;

      .send-v2__asset-dropdown__asset {
        padding: 8px;
      }
    }

    &__input-wrapper {
      border: 1px solid var(--color-border-muted);
      border-radius: 4px;
      height: 100%;

      &--opened {
        position: relative;
        z-index: 2050;
      }

      .send-v2__asset-dropdown__asset {
        height: 100%;

        &:hover {
          background-color: var(--color-background-default);
        }
      }
    }

    &__input {
      @include Paragraph;

      z-index: 1025;
      position: relative;
      height: 54px;
      width: 100%;
      border: none;
      border-radius: 4px;
      background-color: var(--color-background-default);
      color: var(--color-text-alternative);
      padding: 10px;
    }
  }

  &__memo-text-area,
  &__hex-data {
    &__input {
      @include Paragraph;

      z-index: 1025;
      position: relative;
      height: 54px;
      width: 100%;
      border: 1px solid var(--color-border-muted);
      border-radius: 4px;
      background-color: var(--color-background-default);
      color: var(--color-text-alternative);
      padding: 10px;
    }
  }

  &__amount-max {
    @include H7;

    position: relative;
    display: inline-block;
    width: 56px;
    height: 20px;
    margin-top: 5px;
    border-radius: 100px;

    &__button {
      width: 56px;
      height: 20px;
      position: absolute;
      border: 1px solid var(--color-primary-default);
      border-radius: 100px;
      cursor: pointer;
      top: 0;
      left: 0;
      right: 0;
      bottom: 0;
      display: flex;
      align-items: center;
      justify-content: center;
      color: var(--color-primary-default);
      background: var(--color-background-default);

      &__disabled {
        opacity: 0.5;
        cursor: auto;
      }
    }

    input:checked + &__button {
      background-color: var(--color-primary-default);
      border: 1px solid var(--color-primary-default);
      color: var(--color-primary-inverse);
    }
  }

  &__amount-max input {
    opacity: 0;
    width: 0;
    height: 0;
  }

  &__sliders-icon-container {
    @include Paragraph;

    display: flex;
    align-items: center;
    justify-content: center;
    height: 24px;
    width: 24px;
    border: 1px solid var(--color-primary-default);
    border-radius: 4px;
    background-color: var(--color-background-default);
    position: absolute;
    right: 15px;
    top: 14px;
    cursor: pointer;
  }

  &__sliders-icon {
    color: var(--color-primary-default);
  }

  &__memo-text-area {
    &__input {
      padding: 6px 10px;
    }
  }

  &__footer {
    height: 92px;
    width: 100%;
    display: flex;
    justify-content: space-evenly;
    align-items: center;
    border-top: 1px solid var(--color-border-muted);
    background: var(--color-background-default);
    padding: 0 12px;
    flex-shrink: 0;
  }

  &__next-btn,
  &__cancel-btn {
    width: 163px;
    margin: 0 4px;
  }

  &__gas-modal-card {
    width: 360px;
    display: flex;
    flex-flow: column;
    align-items: flex-start;
    padding-left: 20px;

    &__title {
      @include H4;

      height: 26px;
      color: var(--color-text-alternative);
      margin-top: 17px;
    }

    &__copy {
      @include Paragraph;

      height: 38px;
      width: 314px;
      color: var(--color-text-alternative);
      margin-top: 17px;
    }

    .gas-tooltip-input-arrows {
      @include H4;

      width: 32px;
      height: 54px;
      border-left: 1px solid var(--color-border-muted);
      color: var(--color-text-alternative);
      right: 0;
      padding: 0;
      display: flex;
      justify-content: space-around;
      align-items: center;
    }

    .gas-tooltip-input-arrow-wrapper {
      align-items: center;
      align-self: stretch;
      display: flex;
      flex-direction: column;
      flex-grow: 1;
      justify-content: center;
    }

    input[type='number']::-webkit-inner-spin-button {
      -webkit-appearance: none;
      display: none;
    }

    input[type='number']:hover::-webkit-inner-spin-button {
      -webkit-appearance: none;
      display: none;
    }
  }
}

.advanced-gas-options-btn {
  @include H6;

  display: flex;
  justify-content: flex-end;
  color: var(--color-primary-default);
  cursor: pointer;
  margin-top: 5px;
}

.sliders-icon-container {
  @include Paragraph;

  display: flex;
  align-items: center;
  justify-content: center;
  height: 24px;
  width: 24px;
  border: 1px solid var(--color-primary-default);
  border-radius: 4px;
  background-color: var(--color-background-default);
  position: absolute;
  right: 15px;
  top: 14px;
  cursor: pointer;
}

.gas-fee-reset {
  @include H6;

  display: flex;
  align-items: center;
  justify-content: center;
  height: 24px;
  border-radius: 4px;
  background-color: var(--color-background-default);
  position: absolute;
  right: 12px;
  top: 14px;
  cursor: pointer;
  color: var(--color-primary-default);
}

.sliders-icon {
  color: var(--color-primary-default);
}
