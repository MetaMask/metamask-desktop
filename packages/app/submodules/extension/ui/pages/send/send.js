import React, { useEffect, useCallback, useContext, useRef } from 'react';
import { useDispatch, useSelector } from 'react-redux';
import { useHistory, useLocation } from 'react-router-dom';
import {
  addHistoryEntry,
  getDraftTransactionExists,
  getIsUsingMyAccountForRecipientSearch,
  getRecipient,
  getRecipientUserInput,
  getSendStage,
  resetRecipientInput,
  resetSendState,
  SEND_STAGES,
  startNewDraftTransaction,
  updateRecipient,
  updateRecipientUserInput,
} from '../../ducks/send';
import { isCustomPriceExcessive } from '../../selectors';
import { getSendHexDataFeatureFlagState } from '../../ducks/metamask/metamask';
import { showQrScanner } from '../../store/actions';
import { MetaMetricsContext } from '../../contexts/metametrics';
import { EVENT } from '../../../shared/constants/metametrics';
import { ASSET_TYPES } from '../../../shared/constants/transaction';
import SendHeader from './send-header';
import AddRecipient from './send-content/add-recipient';
import SendContent from './send-content';
import SendFooter from './send-footer';
import EnsInput from './send-content/add-recipient/ens-input';

const sendSliceIsCustomPriceExcessive = (state) =>
  isCustomPriceExcessive(state, true);

export default function SendTransactionScreen() {
  const history = useHistory();
  const startedNewDraftTransaction = useRef(false);
  const stage = useSelector(getSendStage);
  const gasIsExcessive = useSelector(sendSliceIsCustomPriceExcessive);
  const isUsingMyAccountsForRecipientSearch = useSelector(
    getIsUsingMyAccountForRecipientSearch,
  );
  const recipient = useSelector(getRecipient);
  const showHexData = useSelector(getSendHexDataFeatureFlagState);
  const userInput = useSelector(getRecipientUserInput);
  const draftTransactionExists = useSelector(getDraftTransactionExists);
  const location = useLocation();
  const trackEvent = useContext(MetaMetricsContext);

  const dispatch = useDispatch();

  const cleanup = useCallback(() => {
    dispatch(resetSendState());
  }, [dispatch]);

  /**
   * It is possible to route to this page directly, either by typing in the url
   * or by clicking the browser back button after progressing to the confirm
   * screen. In the case where a draft transaction does not yet exist, this
   * hook is responsible for creating it. We will assume that this is a native
   * asset send.
   */
  useEffect(() => {
    if (
      draftTransactionExists === false &&
      startedNewDraftTransaction.current === false
    ) {
      startedNewDraftTransaction.current = true;
      dispatch(startNewDraftTransaction({ type: ASSET_TYPES.NATIVE }));
    }
  }, [draftTransactionExists, dispatch]);

  useEffect(() => {
    window.addEventListener('beforeunload', cleanup);
  }, [cleanup]);

  useEffect(() => {
    if (location.search === '?scan=true') {
      dispatch(showQrScanner());

      // Clear the queryString param after showing the modal
      const cleanUrl = window.location.href.split('?')[0];
      window.history.pushState({}, null, `${cleanUrl}`);
      window.location.hash = '#send';
    }
  }, [location, dispatch]);

  useEffect(() => {
    return () => {
      dispatch(resetSendState());
      window.removeEventListener('beforeunload', cleanup);
    };
  }, [dispatch, cleanup]);

  let content;

  if (
    draftTransactionExists &&
    [SEND_STAGES.EDIT, SEND_STAGES.DRAFT].includes(stage)
  ) {
    content = (
      <>
        <SendContent
          showHexData={showHexData}
          gasIsExcessive={gasIsExcessive}
        />
        <SendFooter key="send-footer" history={history} />
      </>
    );
  } else {
    content = <AddRecipient />;
  }

  return (
    <div className="page-container">
      <SendHeader history={history} />
      <EnsInput
        userInput={userInput}
        className="send__to-row"
        onChange={(address) => dispatch(updateRecipientUserInput(address))}
        onValidAddressTyped={async (address) => {
          dispatch(
            addHistoryEntry(`sendFlow - Valid address typed ${address}`),
          );
          await dispatch(updateRecipientUserInput(address));
          dispatch(updateRecipient({ address, nickname: '' }));
        }}
        internalSearch={isUsingMyAccountsForRecipientSearch}
        selectedAddress={recipient.address}
        selectedName={recipient.nickname}
        onPaste={(text) => {
          dispatch(
            addHistoryEntry(
              `sendFlow - User pasted ${text} into address field`,
            ),
          );
        }}
        onReset={() => dispatch(resetRecipientInput())}
        scanQrCode={() => {
          trackEvent({
            event: 'Used QR scanner',
            category: EVENT.CATEGORIES.TRANSACTIONS,
            properties: {
              action: 'Edit Screen',
              legacy_event: true,
            },
          });
          dispatch(showQrScanner());
        }}
      />
      {content}
    </div>
  );
}
