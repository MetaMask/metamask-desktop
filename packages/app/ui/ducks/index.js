import { combineReducers } from 'redux';
import { persistReducer } from 'redux-persist';

// import createElectronStorage from '../helpers/utils/electron-storage';
// import { persistedUIStoreKey } from '../helpers/constants/storage';
import appReducer from './app/app';
import localesReducer from './locale/locale';
import pairStatusReducer from './pair-status/pair-status';

const pairStatusPersistConfig = {
  key: 'pairStatus',
  // storage: createElectronStorage({
  //   name: 'mmd-desktop-ui-v0.0.0-pair-status',
  // }),
  storage: window.electron.pairStatusStore,
  blacklist: ['connections', 'isWebSocketConnected'],
  whitelist: ['isDesktopEnabled', 'isSuccessfulPairSeen', 'lastActivation'],
};
const persistedPairStatusReducer = persistReducer(
  pairStatusPersistConfig,
  pairStatusReducer,
);

const rootReducer = combineReducers({
  app: appReducer,
  locales: localesReducer,
  pairStatus: persistedPairStatusReducer,
});

const rootPersistConfig = {
  key: 'root',
  // storage: createElectronStorage({
  //   name: persistedUIStoreKey,
  // }),
  storage: window.electron.rootStore,
  whitelist: ['app', 'locales'],
};
const persistedRootReducer = persistReducer(rootPersistConfig, rootReducer);

export default persistedRootReducer;
