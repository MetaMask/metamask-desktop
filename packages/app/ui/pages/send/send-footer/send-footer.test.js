import React from 'react';
import configureMockStore from 'redux-mock-store';
import { fireEvent, waitFor } from '@testing-library/react';
import thunk from 'redux-thunk';
import { renderWithProvider } from '../../../../test/lib/render-helpers';
import mockState from '../../../../test/data/mock-state.json';
import {
  CONFIRM_TRANSACTION_ROUTE,
  DEFAULT_ROUTE,
} from '../../../helpers/constants/routes';
import { SEND_STAGES } from '../../../ducks/send';
import SendFooter from '.';

const mockResetSendState = jest.fn();
const mockSendTransaction = jest.fn();
const mockAddtoAddressBook = jest.fn();
const mockCancelTx = jest.fn();

jest.mock('../../../ducks/send/index.js', () => ({
  ...jest.requireActual('../../../ducks/send/index.js'),
  signTransaction: () => mockSendTransaction,
  resetSendState: () => mockResetSendState,
}));

jest.mock('../../../store/actions.js', () => ({
  addToAddressBook: () => mockAddtoAddressBook,
  cancelTx: () => mockCancelTx,
}));

describe('SendFooter Component', () => {
  const props = {
    history: {
      push: jest.fn(),
    },
  };

  afterEach(() => {
    props.history.push.mockReset();
  });

  const mockStore = configureMockStore([thunk])(mockState);

  it('should match snapshot', () => {
    const { container } = renderWithProvider(
      <SendFooter {...props} />,
      mockStore,
    );

    expect(container).toMatchSnapshot();
  });

  describe('onCancel', () => {
    it('should call reset send state and route to recent page without cancelling tx', () => {
      const { queryByText } = renderWithProvider(
        <SendFooter {...props} />,
        mockStore,
      );

      const cancelText = queryByText('Cancel');
      fireEvent.click(cancelText);

      expect(mockResetSendState).toHaveBeenCalled();
      expect(mockCancelTx).not.toHaveBeenCalled();
      expect(props.history.push).toHaveBeenCalledWith(
        '/mostRecentOverviewPage',
      );
    });

    it('should reject/cancel tx when coming from tx editing and route to index', () => {
      const sendDataState = {
        ...mockState,
        send: {
          currentTransactionUUID: '01',
          draftTransactions: {
            '01': {
              id: '99',
            },
          },
          stage: SEND_STAGES.EDIT,
        },
      };

      const sendStateStore = configureMockStore([thunk])(sendDataState);

      const { queryByText } = renderWithProvider(
        <SendFooter {...props} />,
        sendStateStore,
      );

      const rejectText = queryByText('Reject');
      fireEvent.click(rejectText);

      expect(mockResetSendState).toHaveBeenCalled();
      expect(mockCancelTx).toHaveBeenCalled();
      expect(props.history.push).toHaveBeenCalledWith(DEFAULT_ROUTE);
    });
  });

  describe('onSubmit', () => {
    it('should', async () => {
      const { queryByText } = renderWithProvider(
        <SendFooter {...props} />,
        mockStore,
      );

      const nextText = queryByText('Next');
      fireEvent.click(nextText);

      await waitFor(() => {
        expect(mockAddtoAddressBook).toHaveBeenCalled();
        expect(mockSendTransaction).toHaveBeenCalled();
        expect(props.history.push).toHaveBeenCalledWith(
          CONFIRM_TRANSACTION_ROUTE,
        );
      });
    });
  });

  describe('Component Update', () => {
    it('should match snapshot when component updated with errors', () => {
      const { container, rerender } = renderWithProvider(
        <SendFooter.WrappedComponent />,
      );

      const sendErrorProps = {
        sendErrors: {
          gasFee: 'gas fee error',
          amount: 'amount error',
        },
      };

      rerender(<SendFooter.WrappedComponent {...sendErrorProps} />);
      expect(container).toMatchSnapshot();
    });
  });
});
