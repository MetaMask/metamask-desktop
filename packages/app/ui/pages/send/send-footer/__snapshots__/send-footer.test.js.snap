// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SendFooter Component Component Update should match snapshot when component updated with errors 1`] = `
<div>
  <div
    class="page-container__footer"
  >
    <footer>
      <button
        class="button btn--rounded btn-secondary page-container__footer-button"
        data-testid="page-container-footer-cancel"
        role="button"
        tabindex="0"
      >
        [cancel]
      </button>
      <button
        class="button btn--rounded btn-primary page-container__footer-button"
        data-testid="page-container-footer-next"
        role="button"
        tabindex="0"
      >
        [next]
      </button>
    </footer>
  </div>
</div>
`;

exports[`SendFooter Component should match snapshot 1`] = `
<div>
  <div
    class="page-container__footer"
  >
    <footer>
      <button
        class="button btn--rounded btn-secondary page-container__footer-button"
        data-testid="page-container-footer-cancel"
        role="button"
        tabindex="0"
      >
        Cancel
      </button>
      <button
        class="button btn--rounded btn-primary page-container__footer-button"
        data-testid="page-container-footer-next"
        role="button"
        tabindex="0"
      >
        Next
      </button>
    </footer>
  </div>
</div>
`;
